power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +366,550 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+
+    constexpr auto index() const { return index_; }
+    constexpr auto topology_index() const { return topology_index_; }
+
+    IntS id() const {
+        return apply([](auto const& t) { return t.id(); });
+    }
+    IntS tap_pos() const {
+        return apply([](auto const& t) { return t.tap_pos(); });
+    }
+    IntS tap_min() const {
+        return apply([](auto const& t) { return t.tap_min(); });
+    }
+    IntS tap_max() const {
+        return apply([](auto const& t) { return t.tap_max(); });
+    }
+    bool connected_at_tap_side() const {
+        return apply([](auto const& t) { return t.status(t.tap_side()); });
+    }
+    bool connected_at_control_side(TransformerTapRegulator const& regulator) const {
+        return apply([side = regulator.control_side()](auto const& t) {
+            using SideType = typename std::remove_cvref_t<decltype(t)>::SideType;
+            return t.status(static_cast<SideType>(side));
+        });
+    }
+
+    template <typename Func>
+        requires(std::invocable<Func, TransformerTypes const&> && ...)
+    auto apply(Func const& func) const {
+        return std::visit([&func](auto const& t) { return func(t.get()); }, transformer_);
+    }
+
+  private:
+    std::variant<std::reference_wrapper<const TransformerTypes>...> transformer_;
+
+    Idx2D index_;
+    Idx topology_index_;
+};
+
+template <transformer_c... TransformerTypes> struct TapRegulatorRef {
+    std::reference_wrapper<const TransformerTapRegulator> regulator;
+    TransformerWrapper<TransformerTypes...> transformer;
+};
+
+template <typename... Ts> struct transformer_types_s;
+template <> struct transformer_types_s<std::tuple<>> {
+    using type = std::tuple<>;
+};
+template <typename T, typename... Ts> struct transformer_types_s<T, std::tuple<Ts...>> {
+    using type = std::tuple<T, Ts...>;
+};
+template <typename T, typename... Ts> struct transformer_types_s<std::tuple<T, Ts...>> {
+    using type =
+        std::conditional_t<transformer_c<T>,
+                           typename transformer_types_s<T, typename transformer_types_s<std::tuple<Ts...>>::type>::type,
+                           typename transformer_types_s<std::tuple<Ts...>>::type>;
+};
+template <typename... Ts> struct transformer_types_s<std::tuple<std::tuple<Ts...>>> {
+    using type = typename transformer_types_s<std::tuple<Ts...>>::type;
+};
+
+template <typename... Ts> using transformer_types_t = typename transformer_types_s<std::tuple<Ts...>>::type;
+
+template <typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, TransformerTapRegulator>
+TransformerTapRegulator const& find_regulator(State const& state, ID regulated_object) {
+    auto const regulators = get_component_citer<TransformerTapRegulator>(state);
+
+    auto result_it = std::ranges::find_if(regulators, [regulated_object](auto const& regulator) {
+        return regulator.regulated_object() == regulated_object;
+    });
```

Comment:
your comment does not explicitly provide a mapping from one to the other but if i assume correctly, then yes: there can be zero or one regulator of the same type mapped to a regulated object. otherwise, there will be conflicting inputs and you might end up in an infinite loop

NOTE: the fact that there cannot be multiple regulators mapped to the same transformer should be checked by `main_model` (i will add that to the Jira ticket). ~in the optimizer, my guess is that it can be assumed to be undefined behavior~ EDIT: due to the severity of inifinite loops, we cannot make it UB and we indeed need to check here.

Reply:
i changed my mind: this is 100% the transformer ranker's responsibility